asapdiscovery-modeling/asapdiscovery/modeling/protein_prep_v2.py

```diff
@@ -115,28 +154,50 @@ def _prep(self, inputs: list[Complex]) -> list[PreppedComplex]:
-                # check matching du exists
-                du_name = complex.target.target_name + ".oedu"
-                du_path = self.du_cache / du_name
-                if du_path.exists():
-                    prepped_target = PreppedTarget.from_oedu_file(
-                        du_path,
-                        ids=complex.target.ids,
-                        target_name=complex.target.target_name,
-                        ligand_chain=complex.ligand_chain,
-                    )
-                    pc = PreppedComplex(target=prepped_target, ligand=complex.ligand)
-                    prepped_complexes.append(pc)
-                else:
-                    if self.fail_missing_cache:
-                        raise FileNotFoundError(
-                            f"Missing cached design unit: {du_path}"
+        if self.cache_dir:
+            if self.cache_type == CacheType.JSON:
+                for complex in inputs:
+                    json_name = complex.target.target_name + ".json"
+                    json_path = self.cache_dir / json_name
+                    if json_path.exists():
+                        prepped_complexes.append(
+                            PreppedComplex.from_json_file(json_path)
                         )
                     else:
-                        warnings.warn(f"Missing cached design unit: {du_path}")
-                        prepped_complexes.append(None)
+                        if self.fail_missing_cache:
+                            raise FileNotFoundError(f"Missing cached json: {json_path}")
+                        else:
+                            warnings.warn(f"Missing cached json: {json_path}")
+                            prepped_complexes.append(None)
+
+            elif self.cache_type == CacheType.DesignUnit:
+                for complex in inputs:
+                    # check matching du exists
+                    du_name = complex.target.target_name + ".oedu"
+                    du_path = self.cache_dir / du_name
+                    if du_path.exists():
+                        prepped_target = PreppedTarget.from_oedu_file(
+                            du_path,
+                            ids=complex.target.ids,
+                            target_name=complex.target.target_name,
+                            ligand_chain=complex.ligand_chain,
+                        )
+                        pc = PreppedComplex(
+                            target=prepped_target, ligand=complex.ligand
+                        )
+                        prepped_complexes.append(pc)
+                    else:
+                        if self.fail_missing_cache:
+                            raise FileNotFoundError(
+                                f"Missing cached design unit: {du_path}"
+                            )
+                        else:
+                            warnings.warn(f"Missing cached design unit: {du_path}")
+                            prepped_complexes.append(None)
```

Comment:
I wonder if all this logic can be abstracted into its own function i.e. `load_from_cache` or something

Reply:
Done now